### PR TITLE
Use "mqtt" as WS protocol unless using v3

### LINKED ISSUE #408: mqtt over websockets using google chrome
Hello,

I am using this MQTT client in a web application to communicate with a python broker.

In the connect message, the browser (Chrome) outputs in the console the following error:

`WebSocket connection to 'ws://127.0.0.1:8888/' failed: Error during WebSocket handshake: Sent non-empty 'Sec-WebSocket-Protocol' header but no response was received`

In my broker, the following error is presented:

`[2016-05-19 11:46:57,027] :: INFO :: hbmqtt.broker :: Connection from 127.0.0.1:53516 on listener 'ws-mqtt'
[2016-05-19 11:46:57,027] :: INFO :: websockets.protocol :: Failing the WebSocket connection: 1006  
WARNING :: hbmqtt.broker :: MQTT-3.1.0-1: Can't read first packet an CONNECT: No more data`

However, using Firefox, as well as Safari, the MQTT client works fine.

Broker Output for Firefox:

`[2016-05-19 11:47:32,843] :: INFO :: hbmqtt.broker :: Connection from 127.0.0.1:53524 on listener 'ws-mqtt'`

I have a test [project](https://github.com/vasco-santos/mqttOverWebSockets) for this issue.

### PR DESCRIPTION
**Pull request by RangerMauve**
Fixes #408

### CODE REVIEW

**Comment by mcollina** [comment]
Can we add a unit test for this?

**Comment by RangerMauve** [comment]
I'm having a bit of an issue trying to figure out how to check that the protocol was being passed correctly to the websocket. By the looks of it, it should be accessible via `client.stream.socket.protocol`. But it seems like [that only gets set by the server](https://github.com/websockets/ws/blob/f30da5dad547595c7df134be97be919b4cb5062a/lib/WebSocket.js#L552)

**Comment by RangerMauve** [comment]
Alternatively I'd need to somehow set up servers that expect 3.1 or 3.1.1 for the tests, and I think that's a bit outside the range of what I can do at the moment.

**Comment by mcollina** [comment]
@RangerMauve yes, it's get set by the server. Here is the server https://github.com/mqttjs/MQTT.js/blob/master/test/websocket_client.js#L22-L29. you can just create a test and check if the ws connection has the expected subprotocol.

**Comment by RangerMauve** [comment]
What should I do about the encrypted sockets? I have a test over [here](https://github.com/mqttjs/MQTT.js/blob/master/test/abstract_client.js#L256) at the moment that works on non encrypted sockets properly

**Comment by mcollina** [comment]
Don't worry about encrypted websockets.

**Comment by RangerMauve** [comment]
Well, currently I have the test in abstract_client, so it's getting executed for encrypted sockets as well. Is there a better place for me to put it?

**Comment by mcollina** [comment]
Can you please commit those, I don't understand. This is a test specific
for websockets.
Il giorno gio 23 giu 2016 alle 23:52 RangerMauve <notifications@github.com>
ha scritto:

**Comment by RangerMauve** [comment]
Here's my attempt at the tests. It just checks to see that the protocol gets set, to the default, and to `mqttv3.1` if 3.1 is used as you described in the issue.

File: test/abstract_client.js [hunk 1]
```diff
@@ -253,6 +253,28 @@ module.exports = function (server, config) {
...
+      server.once('client', function (client) {
+        client.stream.socket.protocol.should.equal('mqttv3.1');
+        done();
+        client.disconnect();
+      });
+      connect({
+        protocolId: 'MQIsdp',
+        protocolVersion: 3
+      });
+    });
```
Comment (mcollina) [inline test/abstract_client.js]: You should move these to https://github.com/mqttjs/MQTT.js/blob/master/test/websocket_client.js.

**Comment by mcollina** [comment]
Perfect, thanks!

**Comment by mcollina** [comment]
Released as 1.12.0.

**Comment by vasco-santos** [comment]
Hello
First of all, thanks for the update. However, I tested this new version (1.12.0) using Google Chrome and the browser does not connect with the broker yet.

Here is my test project, with a python3 broker and a web page:
[Test Project](https://github.com/vasco-santos/mqttOverWebSockets)

Thanks for your help!

**Pull request merged**